### PR TITLE
Add JAX implementation fol `MatrixIsPositiveDefinite` `Op`

### LINKED ISSUE #6849: ENH: JAX OP `MatrixIsPositiveDefinite`
Hi! I tried to replicate the Correlated priors with `LKJCorr` from https://tomicapretto.github.io/posts/2022-06-12_lkj-prior/#model-4-correlated-priors-with-lkjcorr.-replicate-rstanarm-prior using the `numpyro` sampler and got the error:

### Before

```python
`NotImplementedError: No JAX conversion for the given `Op`: MatrixIsPositiveDefinite`. 
```

So it would be great to add this operator. 

Is this something a newbie like me could try?

### After

```python
We should be able to run the model above wit the JAX sampler from `numpyro`.
```

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes: https://github.com/pymc-devs/pymc/issues/6849

Following https://pytensor.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html

----
:books: Documentation preview :books:: https://pymc--6853.org.readthedocs.build/en/6853/

### CODE REVIEW

File: tests/sampling/test_jax.py [hunk 1]
```diff
@@ -49,6 +52,14 @@ def test_old_import_route():
     assert set(new_sj.__all__) <= set(dir(old_sj))
 
 
+def test_jax_PosDefMatrix():
+    x = pt.tensor()
+    matrix_pos_def = PosDefMatrix()
+    x_is_pos_def = matrix_pos_def(x)
+    out_fg = FunctionGraph(outputs=[x_is_pos_def])
+    compare_jax_and_py(fgraph=out_fg, test_inputs=pt.eye(3))
```
Comment (juanitorduz) [inline tests/sampling/test_jax.py]: @ricardoV94 I can not use `from tests.link.jax.test_basic import compare_jax_and_py` from `pymc` (because it belongs to the tests) ...shall we add a copy?

File: pymc/sampling/jax.py [hunk 2]
```diff
@@ -74,6 +74,15 @@ def assert_fn(value, *inps):
     return assert_fn
 
 
+@jax_funcify.register(PosDefMatrix)
+def jax_funcify_PosDefMatrix(op, **kwargs):
+    def posdefmatrix_fn(value, *inps):
+        no_pos_def = jnp.any(jnp.isnan(jnp.linalg.cholesky(value)))
+        return (no_pos_def == jnp.array(False, dtype="bool")).astype(jnp.int_)
```
Comment (juanitorduz) [inline pymc/sampling/jax.py]: Do we have a better way to negate a statement and convert it to `int`? This smells bad ...

File: tests/sampling/test_jax.py (hunk 1 above)
Comment (ricardoV94) [reply tests/sampling/test_jax.py]: I don't think there's anything special about that. I would just compile a function to JAX and confirm it returns the expected results

Comment (juanitorduz) [reply tests/sampling/test_jax.py]: I want to get the test working with a simple case, and then I will add parametrize it with edge cases.

Comment (ricardoV94) [reply tests/sampling/test_jax.py]: We should compile only once and test with multiple values (pos and not pos). Parametrize is not ideal because we don't want to recompile/jit just to test a new value

File: pymc/sampling/jax.py (hunk 2 above)
Comment (ricardoV94) [reply pymc/sampling/jax.py]: The PyTensor Op doesn't return a bool? It seems like it should

Comment (juanitorduz) [reply pymc/sampling/jax.py]: It returns an integer:

```python
import pytensor.tensor as pt
from pymc.distributions.multivariate import PosDefMatrix

matrix_pos_def = PosDefMatrix()

matrix_pos_def(pt.eye(2)).eval()
>> array(1, dtype=int8)
```

Comment (ricardoV94) [reply pymc/sampling/jax.py]: Let's check if we can change it to bool. Should need tweaking of `make_node` and `perform` of the original Op

Comment (juanitorduz) [reply pymc/sampling/jax.py]: Tried to do it in https://github.com/pymc-devs/pymc/pull/6853/commits/013ead3c8b00b7331429f7702a90092ea733e1a7

Comment (juanitorduz) [reply pymc/sampling/jax.py]: (and tests https://github.com/pymc-devs/pymc/pull/6853/commits/75ca38bdc7f1bf4acee903cfec8b79de3bfa5f10)

Comment (ricardoV94) [reply pymc/sampling/jax.py]: Then the jax one can/shoud return the boolean variable `not any(...)`

Comment (juanitorduz) [reply pymc/sampling/jax.py]: In https://github.com/pymc-devs/pymc/pull/6853/commits/b9a2ef8007bfcce9c166858addf909b3c3bce878 I made the output boolean (as in the tests)

File: pymc/sampling/jax.py [hunk 3]
```diff
@@ -74,6 +74,15 @@ def assert_fn(value, *inps):
     return assert_fn
 
 
+@jax_funcify.register(PosDefMatrix)
+def jax_funcify_PosDefMatrix(op, **kwargs):
+    def posdefmatrix_fn(value, *inps):
+        no_pos_def = jnp.any(jnp.isnan(jnp.linalg.cholesky(value)))
+        return no_pos_def == jnp.array(False, dtype="bool")
```
Comment (ricardoV94) [inline pymc/sampling/jax.py]: Isn't this enough?

```suggestion
        return not no_pos_def
```

Comment (ricardoV94) [reply pymc/sampling/jax.py]: I would make it a one liner because the double negation with the name is confusing

**Comment by ricardoV94** [comment]
We have to check that HSGP failing test

**Comment by juanitorduz** [comment]
> We have to check that HSGP failing test

It is passing for me locally 🤔 
```python
tests/gp/test_hsgp_approx.py .. 
```

File: tests/sampling/test_jax.py [hunk 4]
```diff
@@ -49,6 +51,26 @@ def test_old_import_route():
...
+    argvalues=[
+        (jnp.eye(3), True),
+        (jnp.zeros(shape=(10, 10)), False),
+        (jnp.array([[1, -1.5], [0, 1.2]], dtype="float32"), True),
+        (-1 * jnp.array([[1, -1.5], [0, 1.2]], dtype="float32"), False),
+        (jnp.array([[1, -1.5], [0, -1.2]], dtype="float32"), False),
+        (jnp.array([[1, 0.99, 1], [0.99, 1, 0.999], [1, 0.999, 1]], dtype="float32"), False),
+    ],
+)
+def test_jax_PosDefMatrix(input, expected):
```
Comment (ricardoV94) [inline tests/sampling/test_jax.py]: Let's not parametrize, it's quite wasteful to compile a function for each different value we want to test.

Comment (juanitorduz) [reply tests/sampling/test_jax.py]: So should l loop over the cases inside the test to test all the cases? (after the compilation)

**Comment by juanitorduz** [comment]
> We have to check that HSGP failing test

The tests pass locally 🤔

File: tests/sampling/test_jax.py (hunk 4 above)
Comment (ricardoV94) [reply tests/sampling/test_jax.py]: Yup

**Review by ricardoV94 — APPROVED** [review]

**Comment by juanitorduz** [comment]
Some random (unrelated) tests failed 🤷

**Comment by ricardoV94** [comment]
I am not sure is random, doesn't it use the Op we modified?

**Comment by juanitorduz** [comment]
Well, before the last commit where I simply changed the test parameterization all the tests passed. Can you maybe rerun the failed jobs? I can double check from my side.

**Comment by ricardoV94** [comment]
I don't know if the test is deterministic, so passing once isn't a conclusive thing. I can have a look on Monday

**Comment by ricardoV94** [comment]
There was a bug in the MatrixNormal test, it was supposed to try a different draw in the k2_samp test when it failed (10 times), but it was reusing the same draws everytime.

**Comment by ricardoV94** [comment]
I fixed the test and cleaned the PR git history so we can merge without squash. Let me know if it looks correct (and if you need chages be carefull with fetching the remote branch)

**Comment by juanitorduz** [comment]
Great, thank you! Looks great!

File: tests/gp/test_hsgp_approx.py [hunk 5]
```diff
@@ -178,7 +178,7 @@ def test_prior(self, model, cov_func, X1, parameterization):
             gp = pm.gp.Latent(cov_func=cov_func)
             f2 = gp.prior("f2", X=X1)
 
-            idata = pm.sample_prior_predictive(samples=1000)
+            idata = pm.sample_prior_predictive(samples=1000, random_seed=rng)
```
Comment (juanitorduz) [inline tests/gp/test_hsgp_approx.py:181]: 🙏

Comment (ricardoV94) [reply tests/gp/test_hsgp_approx.py:181]: One day we will get this 5 line PR merged. Just you wait :D
